espresso-server/app/src/androidTest/java/io/appium/espressoserver/lib/model/TouchAction.kt

```diff
@@ -0,0 +1,199 @@
+                .withOrigin(origin)
+                .build()
+
+    private val pause: Action
+        get() = ActionBuilder()
+                .withType(PAUSE)
+                .withDuration(0L)
+                .build()
+
+    @Throws(InvalidArgumentException::class)
+    fun toW3CAction(): List<Action> {
+        val w3cActions: List<Action>
+        when (action) {
+            TouchAction.ActionType.MOVE_TO -> w3cActions = listOf(convertMoveTo())
+            TouchAction.ActionType.PRESS -> w3cActions = convertPress(PRESS_DURATION)
+            TouchAction.ActionType.LONG_PRESS -> w3cActions = convertPress(LONG_PRESS_TIMEOUT + TIMEOUT_BUFFER)
+            TouchAction.ActionType.TAP -> w3cActions = convertPress(TAP_TIMEOUT - TIMEOUT_BUFFER)
+            TouchAction.ActionType.RELEASE -> w3cActions = listOf(convertRelease())
+            TouchAction.ActionType.WAIT -> w3cActions = listOf(convertWait())
+            TouchAction.ActionType.CANCEL -> w3cActions = listOf(convertCancel())
+            else -> throw InvalidArgumentException(String.format("Unsupported action type %s", action))
+        }
+
+        // All touch actions map to 3 actions
+        // For multi-touch actions we need each event to happen synchronously with eachother
+
+        // e.g.) if one input calls press (which maps to move + down + wait) and another input is
+        // calling pause (which maps to wait) we need to add two no-ops to the wait event so that it
+        // doesn't prematurely advance to the next action before the 'press' event finishes
+        return padActionsList(w3cActions)
+    }
+
+    private fun convertCancel(): Action {
+        return ActionBuilder()
+                .withType(POINTER_CANCEL)
+                .build()
+    }
+
+    private fun convertWait(): Action {
+        return ActionBuilder()
+                .withType(PAUSE)
+                .build()
+    }
+
+    private fun convertRelease(): Action {
+        return ActionBuilder()
+                .withType(POINTER_UP)
+                .build()
+    }
+
+    private fun convertMoveTo(): Action {
+        return moveTo
+    }
+
+    private fun convertPress(pressDuration: Long?): List<Action> {
+        // Move to spot
+        val moveAction = moveTo
+
+        // Press down
+        val downAction = ActionBuilder()
+                .withType(POINTER_DOWN)
+                .build()
+
+        // Wait for the press duration
+        val waitAction = ActionBuilder()
+                .withType(PAUSE)
+                .withDuration(pressDuration)
+                .build()
+
+        return Arrays.asList(moveAction, downAction, waitAction)
+    }
+
+    // If an action list has fewer than three actions, pad them with 'pauses' of 0 duration
+    private fun padActionsList(actions: List<Action>): List<Action> {
+        // One Jsonwp Touch Action maps to three W3C Actions
+        val W3C_ACTIONS_PER_TOUCH_ACTION = 3
+
+        val paddedActions = ArrayList<Action>(3)
```

Comment:
how about using map?

Reply:
`paddedActions` is data that expects three `Action` objects in order